File: Matchismo/Xamarin/Matchismo/Core/Deck.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Core
{
	public class Deck
	{
		private readonly List<Card> _cards;

		public Deck ()
		{
			_cards = new List<Card> ();
		}

		public void AddCard(Card card, bool atTop = false)
		{
			if (atTop)
				_cards.Insert (0, card);
			else
				_cards.Add (card);
		}

		public Card DrawRandomCard()
		{
			if (!_cards.Any ())
				return null;

			var rand = new Random ();
			var index = rand.Next () % _cards.Count;
			var card = _cards [index];
			_cards.Remove (card);
			return card;
		}
	}
}

File: Matchismo/ObjC/Matchismo/Matchismo/Model/Deck.m
//
//  Deck.m
//  Matchismo
//
//  Created by Bret Koppel on 8/21/14.
//  Copyright (c) 2014 Bret Koppel. All rights reserved.
//

#import "Deck.h"

@interface Deck()
@property (strong, nonatomic) NSMutableArray *cards;
@end

@implementation Deck

- (NSMutableArray *)cards {
    if (!_cards)
        _cards = [[NSMutableArray alloc] init];
    
    return _cards;
}

- (void)addCard:(Card *)card atTop:(BOOL)atTop {
    if (atTop)
        [self.cards insertObject:card atIndex:0];
    else
        [self.cards addObject:card];
}

- (void)addCard:(Card *)card {
    [self addCard:card atTop:NO];
}

- (Card *)drawRandomCard{
    Card *randomCard = nil;
    
    // calling an empty array with an index will throw so...
    if ([self.cards count]){
        unsigned index = arc4random() % [self.cards count];
        randomCard = self.cards[index];
        [self.cards removeObjectAtIndex:index];
    }
    
    return randomCard;
}

@end
